source/extensions/access_loggers/grpc/tcp_config.cc

```diff
@@ -0,0 +1,51 @@
+#include "extensions/access_loggers/grpc/tcp_config.h"
+
+#include "envoy/config/accesslog/v2/als.pb.validate.h"
+#include "envoy/config/filter/accesslog/v2/accesslog.pb.validate.h"
+#include "envoy/registry/registry.h"
+#include "envoy/server/filter_config.h"
+
+#include "common/common/assert.h"
+#include "common/common/macros.h"
+#include "common/grpc/async_client_impl.h"
+#include "common/protobuf/protobuf.h"
+
+#include "extensions/access_loggers/grpc/config_utils.h"
+#include "extensions/access_loggers/grpc/grpc_access_log_proto_descriptors.h"
+#include "extensions/access_loggers/grpc/tcp_grpc_access_log_impl.h"
+#include "extensions/access_loggers/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace AccessLoggers {
+namespace TcpGrpc {
+
+AccessLog::InstanceSharedPtr
+TcpGrpcAccessLogFactory::createAccessLogInstance(const Protobuf::Message& config,
+                                                 AccessLog::FilterPtr&& filter,
+                                                 Server::Configuration::FactoryContext& context) {
+  GrpcCommon::validateProtoDescriptors();
+
+  const auto& proto_config =
+      MessageUtil::downcastAndValidate<const envoy::config::accesslog::v2::TcpGrpcAccessLogConfig&>(
+          config);
+
+  return std::make_shared<TcpGrpcAccessLog>(std::move(filter), proto_config, context.threadLocal(),
+                                            GrpcCommon::getGrpcAccessLoggerCacheSingleton(context));
+}
+
+ProtobufTypes::MessagePtr TcpGrpcAccessLogFactory::createEmptyConfigProto() {
+  return std::make_unique<envoy::config::accesslog::v2::TcpGrpcAccessLogConfig>();
+}
+
+std::string TcpGrpcAccessLogFactory::name() const { return AccessLogNames::get().TcpGrpc; }
+
+/**
+ * Static registration for the HTTP gRPC access log. @see RegisterFactory.
```

Comment:
Why do we have distinct TCP and HTTP access loggers? Can't we have the same implementation provide both?

Reply:
That is how the config is written initially @mattklein123, and the logging interface is not aware of TCP/HTTP, i.e. if you use current http_grpc logger on tcp_proxy, it would emit partially populated HTTPAccessLogEntry.

I did refactor most of the common part to same implementation so it is mostly sharing them, so this is just two registration with two wrapper.